### DISCUSSION TITLE
Filter and effects in Python

### DISCUSSION THREAD

**Discussion by ExtraMobs** [olive-editor/olive#2250] [Ideas]
An option that allows users to write filters and some effects using Python.

I think of something like:

`def process(numpy_array, current_frame, video):
    ...
`

video variable contains information like; max_seconds, frame_count, fps, etc
this is something i thought about doing but never got around to doing

as python is a popular language, many editors and common people know how to use it, and this encourages people to create what they need and cannot easily find

**Comment by mcDandy** [comment]
I do not thing python would be performent enough (be able to run on GPU since every node runs on GPU and CPU does not have the capability of massive paralelism GPUs have).

**Reply by ExtraMobs** [reply]
While there are ways to create and manipulate data in GPU video memory using pytorch (recent research), the idea was to be affordable, performance-free, easy but not accurate, a tradeoff.

a resource that moves the community and complements the project well, as the user does not need to change editors if he does not have the effect he wants, he can do it himself or find it on the internet in a few lines of code

**Comment by kerrymatten** [comment]
hello . nice job

**Comment by Anyeos** [comment]
I recently just think about an idea of a video editor that used all Python. Because it is possible to scale and modularize each part and only run low level parts on libraries. I think that is a good idea but there are a lot of video editors developed only on C++. I used Python to edit videos manually and it is very fast, it is not slow. Of course, Python is used only to execute instructions like "convert_video_to_grayscale(video)" so that is not really processed in a Python loop that of course will be very slow. 
So what if at least Olive include Python like Blender does for addons. The community will start to create addons and that will be very helpful. Going not too far, OBS have Python support for things like that, so using Python is not an irrational idea.

**Reply by ExtraMobs** [reply]
Thought I'd only use it for video processing, but it would be useful for addons in general, thanks for supporting the idea
